### PR TITLE
Travis: Check lowest dependencies

### PR DESCRIPTION
**Pull request by core23**
Travis should check the lowest dependencies required in the `composer.json` file.

### CODE REVIEW

**Comment by core23** [comment]
Ping @compeak

**Comment by compeak** [comment]
Thank you.

**Comment by compeak** [comment]
@core23: [Apparently there are issues with phpcs.](https://travis-ci.org/heiseonline/shariff-backend-php/jobs/145737902)

File: .travis.yml [hunk 1]
```diff
@@ -7,11 +7,17 @@ php:
...
+  include:
+    - php: 5.5
+      env: COMPOSER_FLAGS="--prefer-lowest"
+
 before_script:
  - bash -c 'if [ -n "$GH_TOKEN" ]; then composer config github-oauth.github.com ${GH_TOKEN}; fi;'
  - composer self-update
- - composer install --prefer-dist --no-interaction
  - composer require "squizlabs/php_codesniffer=*"
+ - travis_wait composer update --prefer-dist --no-interaction $COMPOSER_FLAGS
```
Comment (liayn) [inline .travis.yml]: CI should NEVER run a `composer update`. Actually `update` must only be used by a dev, who upgrades dependencies on purpose

Comment (core23) [inline .travis.yml]: We need an update, otherwise we would install all versions from the composer.lock file

Comment (liayn) [inline .travis.yml]: Well that's exactly what CI should do. Install what is nailed in the lock file

Comment (core23) [inline .travis.yml]: That's not the idea of composer. We can specify a range of versions that will work and we should test these range.

Comment (liayn) [inline .travis.yml]: Where does that statement stem from? IMHO this is just simply totally plain wrong.
If you test a range, you would need to test each and every version in every combination with every composer package you use.
Composer assumes that semver is used, hence if you carefully define your dependencies, no breaking change should occur, no matter which version is used.
The CI environment is not there to test the various dependency variations, but is there to validate that what you tested locally and committed is still valid, even if more/other changes to the code have been made.
Therefore it has to validate against the specific version you used locally. That's the reason why every package manager (composer, gem, ...) has a lock file.
That's also the sole reason why you commit the lock file. In order to have a reproducable state on CI systems.

If your composer library is required in a different project, your project's lock file is ignored anyway and the actually chosen lib versions are persisted in the other project's lock file, again documenting the exact versions of code that has been tested in this very project.

Check also https://getcomposer.org/doc/01-basic-usage.md#composer-lock-the-lock-file for more details.

**Comment by core23** [comment]
> Composer assumes that semver is used...

That's right, but you can't guarantee that every projects respects semver. We could have breaks across minor version, because we are all people.

Lets imagine some other fictional case:
Someone creates a PR which relies on an external `a-library` version 1.1, but the `composer.json` file still has a `"a-library": ^1.0` dependency, because he didn't think about updating the composer definition. The `composer.lock` file would say everything is fine, because it used the latest 1.1 version.
If another person want's to use this library with version 1.0 of `a-library`, everything would look fine, but it isn't.

**Comment by liayn** [comment]
Well, if you remove the semver assumption then the whole process is totally different. Then you should nail the version to the last digit in your composer.json. Anything else would be potentially breaking.
(or at least test a few versions and set a safe range)

Regarding your example: This is exactly what I meant above with "you have to check every combination of library versions then". If you use your CI as a compatibility tester aka "let see in which version this works", then we talk about a different system not a continuous integration workflow.
If ones messes up the dependencies it is always a hard fail. In case of using a new feature and not declaring the necessary version dependency is such a hard fail.

**Comment by core23** [comment]
But testing the lowest possible version (too) against testing only the latest version isn't that much pain.

**Comment by liayn** [comment]
But what's the point of testing the oldest and latest version and not testing the version you used during development? IMO there are only two useful cases: Either test what is locked or test *all* versions, anything else is useless overhead.
But @compeak has to decide that anyway.
The only thing is that unit tests are broken currently, that's a matter of fact. My PR #100 reverts your commit again, such that we have stable tests again.

**Comment by compeak** [comment]
We will test the locked versions. Exactly specifying the dependency versions to use is what the lockfile is for.

**Pull request merged**